ChainmailleDesigner/Features/CommandHistory.cs

```diff
@@ -0,0 +1,129 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using ChainmailleDesigner.Features.CommandHistorySupport;
+
+namespace ChainmailleDesigner.Features
+{
+    /// <summary>
+    /// Architecture inspired by an article from Nasi Jofche
+    /// https://betterprogramming.pub/utilizing-the-command-pattern-to-support-undo-redo-and-history-of-operations-b28fa9d58910
+    /// 
+    /// Undo/Redo system for Chainmaille Designer
+    /// Can handle single items or grouped batches of items as a single entry.
+    /// </summary>
+    public class CommandHistory
+    {
+        public delegate void HistoryChangedEventHandler(object source, EventArgs args);
+        public event HistoryChangedEventHandler HistoryChanged;
+
+        public class HistoryStatus : EventArgs
+        {
+            public bool HasUndoAvailable { get; set; }
+            public bool HasRedoAvailable { get; set; }
+        }
+
+        public static CommandHistory instance { get; private set; }
+
+        private HistoryStack<List<IAction>> stackUndo { get; set; }
+        private HistoryStack<List<IAction>> stackRedo { get; set; }
+
+        public static bool HasUndoAvailable { get { return instance.stackUndo.HasItems; } }
+        public static bool HasRedoAvailable { get { return instance.stackRedo.HasItems; } }
+
+        static CommandHistory()
+        {
+            instance = new CommandHistory();
+        }
+
+        private CommandHistory()
+        {
+            stackUndo = new HistoryStack<List<IAction>>();
+            stackRedo = new HistoryStack<List<IAction>>();
+        }
+
+        /// <summary>
+        /// Record an item that can be undone
+        /// </summary>
+        /// <param name="newAction">An IAction item to be recorded on the undo stack</param>
+        public static void Executed(IAction newAction)
+        {
+            if (newAction != null)
+            {
+                var SingleActionList = new List<IAction>();
+                SingleActionList.Add(newAction);
+                instance.stackUndo.Push(SingleActionList, limitQueue: true);
+                instance.ClearReverse();
+            }
```

Comment:
You need an instance.TriggerEvent() here before exiting the inner block.  Otherwise the undo/redo menu enablements never change from the default of (false, false) when undoable actions are added to the stack.